src/opt/test/test_renderer.cc

```diff
@@ -122,43 +142,51 @@ void TestRendererPixelAccuracy(const Camera& camera) {
+  cv::waitKey(0);
   
   // Compare rendered images to the mesh.
   // NOTE: Due to placing vertices exactly at pixel positions and the fill rule
   // used in rendering, the right- and bottom-most pixels will not be covered by
   // the mesh. Thus the test checks that they are still at their original value.
-  const float* depth_ptr = depth_image.get();
   std::size_t i = 0;
-  for (int y = 0; y < camera.height(); ++ y) {
-    for (int x = 0; x < camera.width(); ++ x) {
+  for (int y = 0; y < camera.height(); y += kStep) {
+    for (int x = 0; x < camera.width(); x += kStep) {
       cv::Vec3b rgb_color = color_image(y, x);
-      if (y < camera.height() - 1 && x < camera.width() - 1) {
-        ASSERT_NEAR(*depth_ptr, mesh_vertex_cloud[i].z, 1e-3f)
-            << "Error at pixel (" << x << ", " << y << ")";
-        ASSERT_EQ(rgb_color[0], mesh_vertex_cloud[i].r)
-            << "Error at pixel (" << x << ", " << y << ")";
-        ASSERT_EQ(rgb_color[1], mesh_vertex_cloud[i].g)
-            << "Error at pixel (" << x << ", " << y << ")";
-        ASSERT_EQ(rgb_color[2], mesh_vertex_cloud[i].b)
-            << "Error at pixel (" << x << ", " << y << ")";
-      } else {
-        ASSERT_EQ(*depth_ptr, 0)
-            << "Error at pixel (" << x << ", " << y << ")";
-        ASSERT_EQ(rgb_color[0], 0)
-            << "Error at pixel (" << x << ", " << y << ")";
-        ASSERT_EQ(rgb_color[1], 0)
-            << "Error at pixel (" << x << ", " << y << ")";
-        ASSERT_EQ(rgb_color[2], 0)
-            << "Error at pixel (" << x << ", " << y << ")";
+      float depth_value = depth_mat(y, x);
+      pcl::PointXYZRGB point = mesh_vertex_cloud[i];
+      Eigen::Vector3f p = point.getVector3fMap();
+      Eigen::Vector2f image_point = camera.ProjectToImageCoordinates(Eigen::Vector2f(p.x()/p.z(), p.y()/p.z()));
+      if(!isinf(image_point.squaredNorm())){
+        if (depth_value > 0) {
+          ASSERT_NEAR(x, image_point.x(), 1e-2f);
+          ASSERT_NEAR(y, image_point.y(), 1e-2f);
+          ASSERT_NEAR(depth_value, point.z, 5e-2f)
+              << "Error at pixel (" << x << ", " << y << ") i=" << i;
+          EXPECT_EQ(rgb_color[0], point.r)
+              << "Error at pixel (" << x << ", " << y << ")";
+          EXPECT_EQ(rgb_color[1], point.g)
+              << "Error at pixel (" << x << ", " << y << ")";
+          EXPECT_EQ(rgb_color[2], point.b)
+              << "Error at pixel (" << x << ", " << y << ")";
+        } else {
+          ASSERT_EQ(depth_value, 0)
+              << "Error at pixel (" << x << ", " << y << ")";
+          ASSERT_EQ(rgb_color[0], 0)
+              << "Error at pixel (" << x << ", " << y << ")";
+          ASSERT_EQ(rgb_color[1], 0)
+              << "Error at pixel (" << x << ", " << y << ")";
+          ASSERT_EQ(rgb_color[2], 0)
+              << "Error at pixel (" << x << ", " << y << ")";
+          ASSERT_TRUE((depth_mat(y+1, x) == 0 || depth_mat(y, x+1) == 0));
```

Comment:
Could this cause out-of-bounds accesses in depth_mat, since no check seems to be done that x resp. y is not on the border?